lib/grpc_reflection/service/builder.ex

```diff
@@ -2,200 +2,147 @@ defmodule GrpcReflection.Service.Builder do
-    # protobuf_generate populates services with file_descriptors
-    case service.descriptor() do
-      %FileDescriptorProto{service: [proto]} ->
-        # we should read and use the file descriptor directly instead
-        # of dropping relevant data and trying to discover it
-        process_service_descriptor(name, proto, syntax)
-
-      %ServiceDescriptorProto{} = proto ->
-        process_service_descriptor(name, proto, syntax)
-    end
-  end
-
-  defp process_service_descriptor(service_name, descriptor, syntax) do
-    referenced_types = Util.types_from_descriptor(descriptor)
-
-    method_symbols =
-      Enum.map(
-        descriptor.method,
-        fn method -> service_name <> "." <> method.name end
-      )
-
-    unencoded_payload = process_common(service_name, descriptor, syntax)
-    payload = FileDescriptorProto.encode(unencoded_payload)
-    response = %{file_descriptor_proto: [payload]}
+    service_name = service.__meta__(:name)
+    service_response = build_response(service_name, service)
 
     State.new()
-    |> State.add_symbols(
-      method_symbols
-      |> Enum.reduce(%{}, fn name, acc -> Map.put(acc, name, response) end)
-      |> Map.put(service_name, response)
-    )
-    |> State.add_files(%{(service_name <> ".proto") => response})
-    |> State.add_references(referenced_types)
+    |> State.add_symbols(%{service_name => service_response})
+    |> State.add_files(%{(service_name <> ".proto") => service_response})
+    |> trace_service_refs(service)
   end
 
-  defp process_reference(symbol) do
-    symbol
-    |> Util.convert_symbol_to_module()
-    |> then(fn mod ->
-      descriptor = mod.descriptor()
-      name = symbol
-
-      nested_types = Util.get_nested_types(name, descriptor)
-
-      referenced_types =
-        Util.types_from_descriptor(descriptor)
-        |> Enum.uniq()
-        |> Kernel.--(nested_types)
-
-      unencoded_payload = process_common(name, descriptor, Util.get_syntax(mod), nested_types)
-      payload = FileDescriptorProto.encode(unencoded_payload)
-      response = %{file_descriptor_proto: [payload]}
-
-      root_symbols = %{symbol => response}
-
-      root_symbols =
-        Enum.reduce(nested_types, root_symbols, fn name, acc -> Map.put(acc, name, response) end)
-
-      root_files = %{(symbol <> ".proto") => response}
-
-      root_files =
-        Enum.reduce(nested_types, root_files, fn name, acc ->
-          Map.put(acc, name <> ".proto", response)
-        end)
-
-      extension_file = symbol <> "Extension.proto"
-
-      {root_extensions, root_files} =
-        process_extensions(mod, symbol, extension_file, descriptor, root_files)
-
-      State.new()
-      |> State.add_files(root_files)
-      |> State.add_symbols(root_symbols)
-      |> State.add_extensions(root_extensions)
-      |> State.add_references(referenced_types)
+  defp trace_service_refs(state, module) do
+    service_name = module.__meta__(:name)
+    methods = module.descriptor().method
```

Comment:
Similar to the `FileDescriptorProto` logic in `build_response/2` below, I believe we need a check here:

```suggestion
    methods =
      case module.descriptor() do
        %FileDescriptorProto{service: [proto]} -> proto.method
        proto -> proto.method
      end
```